v2/frame.js

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var TypedError = require('error/typed');
+var read = require('../lib/read');
+var write = require('../lib/write');
+
+var ExtraFrameDataError = TypedError({
+    type: 'tchannel.extra-frame-data',
+    message: 'got {trailing} bytes of extra frame data beyond body',
+    length: null,
+    trailing: null
+});
+
+var FieldTooLargeError = TypedError({
+    type: 'tchannel.field-too-large',
+    message: '{field} too large at {length} bytes, cannot exceed {max} bytes',
+    field: null,
+    length: null,
+    max: null
+});
+
+var InvalidFrameTypeError = TypedError({
+    type: 'tchannel.invalid-frame-type',
+    message: 'invalid frame type {typeNumber}',
+    typeNumber: null
+});
+
+/* jshint maxparams:5 */
+
+module.exports = Frame;
+
+function Frame(id, body) {
+    if (!(this instanceof Frame)) {
+        return new Frame(id, body);
+    }
+    var self = this;
+    if (id === null || id === undefined) {
+        self.id = Frame.NullId;
+    } else {
+        self.id = id;
+    }
+    self.body = body;
+}
+
+Frame.Overhead = 0x10;
+Frame.MayBodySize = 0xffff - Frame.Overhead;
```

Comment:
yes streaming support will require a whole new layer in the tchannel library as yet unstarted